sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -571,6 +571,7 @@ object TypeCoercion {
         NaNvl(l, Cast(r, DoubleType))
       case NaNvl(l, r) if l.dataType == FloatType && r.dataType == DoubleType =>
         NaNvl(Cast(l, DoubleType), r)
+      case NaNvl(l, r) if r.dataType == NullType => NaNvl(l, Cast(r, l.dataType))
```

Comment:
One question I have is, why `NaNvl(FloatType, DoubleType)` should be cast to `NaNvl(DoubleType, DoubleType)`, but `NaNvl(FloatType, NullType)` should not be cast to `NaNvl(DoubleType, DoubleType)`?

They all change the input type from `FloatType` to `DoubleType`. Won't the first cast cause mismatching?